---
 src/disk.c |    5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

Index: lmbench-2.5/src/disk.c
===================================================================
--- lmbench-2.5.orig/src/disk.c
+++ lmbench-2.5/src/disk.c
@@ -7,6 +7,7 @@
  * Bits of this are derived from work by Ethan Solomita.
  */
 
+#define _LARGEFILE64_SOURCE
 #include	<stdio.h>
 #include	<sys/types.h>
 #include	<unistd.h>
@@ -291,9 +292,7 @@ int
 seekto(int fd, uint64 off)
 {
 #ifdef	__linux__
-	extern	loff_t llseek(int, loff_t, int);
-
-	if (llseek(fd, (loff_t)off, SEEK_SET) == (loff_t)-1) {
+	if (lseek64(fd, (loff_t)off, SEEK_SET) == (loff_t)-1) {
 		return(-1);
 	}
 	return (0);
