diff --git a/libgomp/config/linux/x86/futex.h b/libgomp/config/linux/x86/futex.h
index ead74d1496736a49694ef6b9b2b4da50f9852664..6f7406684f700867ac78a1588a7ede190666ca86 100644
--- a/libgomp/config/linux/x86/futex.h
+++ b/libgomp/config/linux/x86/futex.h
@@ -30,13 +30,15 @@
 #  define SYS_futex	202
 # endif
 
+#include "shim_entry_api.h"
+
 static inline void
 futex_wait (int *addr, int val)
 {
   long res;
 
   register long r10 __asm__("%r10") = 0;
-  __asm volatile ("syscall"
+  __asm volatile ("SYSCALLDB"
 		  : "=a" (res)
 		  : "0" (SYS_futex), "D" (addr), "S" (gomp_futex_wait),
 		    "d" (val), "r" (r10)
@@ -45,7 +47,7 @@ futex_wait (int *addr, int val)
     {
       gomp_futex_wait &= ~FUTEX_PRIVATE_FLAG;
       gomp_futex_wake &= ~FUTEX_PRIVATE_FLAG;
-      __asm volatile ("syscall"
+      __asm volatile ("SYSCALLDB"
 		      : "=a" (res)
 		      : "0" (SYS_futex), "D" (addr), "S" (gomp_futex_wait),
 			"d" (val), "r" (r10)
@@ -58,7 +60,7 @@ futex_wake (int *addr, int count)
 {
   long res;
 
-  __asm volatile ("syscall"
+  __asm volatile ("SYSCALLDB"
 		  : "=a" (res)
 		  : "0" (SYS_futex), "D" (addr), "S" (gomp_futex_wake),
 		    "d" (count)
@@ -67,7 +69,7 @@ futex_wake (int *addr, int count)
     {
       gomp_futex_wait &= ~FUTEX_PRIVATE_FLAG;
       gomp_futex_wake &= ~FUTEX_PRIVATE_FLAG;
-      __asm volatile ("syscall"
+      __asm volatile ("SYSCALLDB"
 		      : "=a" (res)
 		      : "0" (SYS_futex), "D" (addr), "S" (gomp_futex_wake),
 			"d" (count)
diff --git a/libgomp/config/linux/x86/shim_entry_api.h b/libgomp/config/linux/x86/shim_entry_api.h
new file mode 120000
index 0000000000000000000000000000000000000000..3d9b3b4733c21849804515f0dd05d23076f68dcf
--- /dev/null
+++ b/libgomp/config/linux/x86/shim_entry_api.h
@@ -0,0 +1 @@
+../../../../../shim/include/arch/x86_64/shim_entry_api.h
\ No newline at end of file
