diff --git a/libgomp/config/linux/x86/futex.h b/libgomp/config/linux/x86/futex.h
index ead74d1496736a49694ef6b9b2b4da50f9852664..113f7983047fcfe9f24162085251a7148ceca556 100644
--- a/libgomp/config/linux/x86/futex.h
+++ b/libgomp/config/linux/x86/futex.h
@@ -30,13 +30,26 @@
 #  define SYS_futex	202
 # endif
 
+#include "shim_entry_offsets.h"
+
+#define XSTR(s) STR(s)
+#define STR(s) #s
+
+asm (
+     ".macro SYSCALLDB\n"
+     "leaq .Lafter_syscalldb\\@(%rip), %rcx\n"
+     "jmpq *%gs:" XSTR(SHIM_SYSCALLDB_OFFSET) "\n"
+     ".Lafter_syscalldb\\@:\n"
+     ".endm\n"
+);
+
 static inline void
 futex_wait (int *addr, int val)
 {
   long res;
 
   register long r10 __asm__("%r10") = 0;
-  __asm volatile ("syscall"
+  __asm volatile ("SYSCALLDB"
 		  : "=a" (res)
 		  : "0" (SYS_futex), "D" (addr), "S" (gomp_futex_wait),
 		    "d" (val), "r" (r10)
@@ -45,7 +58,7 @@ futex_wait (int *addr, int val)
     {
       gomp_futex_wait &= ~FUTEX_PRIVATE_FLAG;
       gomp_futex_wake &= ~FUTEX_PRIVATE_FLAG;
-      __asm volatile ("syscall"
+      __asm volatile ("SYSCALLDB"
 		      : "=a" (res)
 		      : "0" (SYS_futex), "D" (addr), "S" (gomp_futex_wait),
 			"d" (val), "r" (r10)
@@ -58,7 +71,7 @@ futex_wake (int *addr, int count)
 {
   long res;
 
-  __asm volatile ("syscall"
+  __asm volatile ("SYSCALLDB"
 		  : "=a" (res)
 		  : "0" (SYS_futex), "D" (addr), "S" (gomp_futex_wake),
 		    "d" (count)
@@ -67,7 +80,7 @@ futex_wake (int *addr, int count)
     {
       gomp_futex_wait &= ~FUTEX_PRIVATE_FLAG;
       gomp_futex_wake &= ~FUTEX_PRIVATE_FLAG;
-      __asm volatile ("syscall"
+      __asm volatile ("SYSCALLDB"
 		      : "=a" (res)
 		      : "0" (SYS_futex), "D" (addr), "S" (gomp_futex_wake),
 			"d" (count)
diff --git a/libgomp/config/linux/x86/shim_entry_offsets.h b/libgomp/config/linux/x86/shim_entry_offsets.h
new file mode 120000
index 0000000000000000000000000000000000000000..2dc6a4c9d80b213c70d6523fec7932e3ce6046e7
--- /dev/null
+++ b/libgomp/config/linux/x86/shim_entry_offsets.h
@@ -0,0 +1 @@
+../../../../../shim/include/shim_entry_offsets.h
\ No newline at end of file
