diff --git a/nis/nis_call.c b/nis/nis_call.c
index 3fa37e45a3bc1f4b7539cfab8b895dfb25067dad..2537f2ac6f81a249c57c489bd3143ff9010c1858 100644
--- a/nis/nis_call.c
+++ b/nis/nis_call.c
@@ -679,7 +679,7 @@ nis_server_cache_add (const_nis_name name, int search_parent,
 
   /* Choose which entry should be evicted from the cache.  */
   loc = &nis_server_cache[0];
-  if (*loc != NULL)
+  if (*loc != NULL) {
     for (i = 1; i < 16; ++i)
       if (nis_server_cache[i] == NULL)
 	{
@@ -690,6 +690,7 @@ nis_server_cache_add (const_nis_name name, int search_parent,
 	       || ((*loc)->uses == nis_server_cache[i]->uses
 		   && (*loc)->expires > nis_server_cache[i]->expires))
 	loc = &nis_server_cache[i];
+  }
   old = *loc;
   *loc = new;
 
diff --git a/stdlib/setenv.c b/stdlib/setenv.c
index da61ee0720c84109ad96a81749494810d5b59183..3c24c52d5f0d5fbcc088a1f6716bca8c9dfd7eb4 100644
--- a/stdlib/setenv.c
+++ b/stdlib/setenv.c
@@ -276,7 +276,7 @@ unsetenv (const char *name)
   LOCK;
 
   ep = __environ;
-  if (ep != NULL)
+  if (ep != NULL) {
     while (*ep != NULL)
       if (!strncmp (*ep, name, len) && (*ep)[len] == '=')
 	{
@@ -290,6 +290,7 @@ unsetenv (const char *name)
 	}
       else
 	++ep;
+  }
 
   UNLOCK;
 
