diff --git a/shlib-versions b/shlib-versions
index 78b0ad7..5c3dcf2 100644
--- a/shlib-versions
+++ b/shlib-versions
@@ -64,6 +64,9 @@ sh.*-.*-linux.*		ld=ld-linux.so.2	GLIBC_
 # The -ldl interface (see <dlfcn.h>) is the same on all platforms.
 libdl=2
 
+# Interface for Graphene
+liblibos=1
+
 # So far the -lutil interface is the same on all platforms, except for the
 # `struct utmp' format, which depends on libc.
 libutil=1
