diff --git a/elf/syscallas.S b/elf/syscallas.S
new file mode 120000
index 0000000000000000000000000000000000000000..107bd13a746036488e8bd7d4dc4aee265ceabffa
--- /dev/null
+++ b/elf/syscallas.S
@@ -0,0 +1 @@
+../syscallas.S
\ No newline at end of file
diff --git a/elf/syscalldb.c b/elf/syscalldb.c
new file mode 120000
index 0000000000000000000000000000000000000000..61dca527ee3bd1027a81a28264d82a96faaa76a2
--- /dev/null
+++ b/elf/syscalldb.c
@@ -0,0 +1 @@
+../syscalldb.c
\ No newline at end of file
diff --git a/syscallas.S b/syscallas.S
new file mode 100644
index 0000000000000000000000000000000000000000..07c6ffce0f3b1bc57e5b1ca26086a567dfbf5a41
--- /dev/null
+++ b/syscallas.S
@@ -0,0 +1,11 @@
+#include <syscalldb.h>
+
+.weak syscalldb
+.type syscalldb,@function
+
+syscalldb:
+	.cfi_startproc
+	syscall
+	retq
+	.cfi_endproc
+	.size syscalldb,.-syscalldb
diff --git a/syscalldb.c b/syscalldb.c
new file mode 100644
index 0000000000000000000000000000000000000000..47b17de10f831f28b645686995c5df007bb375d3
--- /dev/null
+++ b/syscalldb.c
@@ -0,0 +1,11 @@
+#include "syscalldb.h"
+#include <stdarg.h>
+#include <errno.h>
+
+int register_library (const char * name, unsigned long load_address)
+	__attribute__((weak));
+
+int register_library (const char * name, unsigned long load_address)
+{
+	return 0;
+}
diff --git a/syscalldb.h b/syscalldb.h
new file mode 100644
index 0000000000000000000000000000000000000000..4914e280d8cd6a62ab1ea4c9aa41601763167e7d
--- /dev/null
+++ b/syscalldb.h
@@ -0,0 +1,37 @@
+#ifndef _SYSCALLDB_H_
+#define _SYSCALLDB_H_
+
+#ifdef __ASSEMBLER__
+.weak syscalldb
+.type syscalldb, @function
+
+# if defined(PSEUDO) && defined(SYSCALL_NAME) && defined(SYSCALL_SYMBOL)
+#  define SYSCALLDB                     \
+    subq $128, %rsp;                    \
+    .cfi_adjust_cfa_offset 128;         \
+    callq *syscalldb@GOTPCREL(%rip);    \
+    addq $128, %rsp;                    \
+    .cfi_adjust_cfa_offset -128
+# else
+#  define SYSCALLDB                             \
+    callq *syscalldb@GOTPCREL(%rip)
+# endif
+
+#else /* !__ASSEMBLER__ */
+asm (
+".weak syscalldb\r\n"
+".type syscalldb, @function\r\n");
+
+#define SYSCALLDB                           \
+    "subq $128, %%rsp\n\t"                  \
+    ".cfi_adjust_cfa_offset 128\n\t"        \
+    "callq *syscalldb@GOTPCREL(%%rip)\n\t"  \
+    "addq $128, %%rsp\n\t"                  \
+    ".cfi_adjust_cfa_offset -128\n\t"
+
+#define SYSCALLDB_ASM                       \
+    "callq *syscalldb@GOTPCREL(%rip)\n\t"
+
+#endif /* Assembler */
+
+#endif /* _SYSCALLDB_H */
diff --git a/sysdeps/unix/sysv/linux/x86_64/syscalldb.h b/sysdeps/unix/sysv/linux/x86_64/syscalldb.h
new file mode 120000
index 0000000000000000000000000000000000000000..55ab305032a38796e4abc8ce7a208bdaefe8639e
--- /dev/null
+++ b/sysdeps/unix/sysv/linux/x86_64/syscalldb.h
@@ -0,0 +1 @@
+../../../../../syscalldb.h
\ No newline at end of file
