diff --git a/shim_entry_offsets.h b/shim_entry_offsets.h
new file mode 120000
index 0000000000000000000000000000000000000000..752b85f99302ddfe1403186f1118b91560255ac2
--- /dev/null
+++ b/shim_entry_offsets.h
@@ -0,0 +1 @@
+../shim/include/shim_entry_offsets.h
\ No newline at end of file
diff --git a/syscalldb.h b/syscalldb.h
new file mode 100644
index 0000000000000000000000000000000000000000..6c1af42bdfc62744c0d9518ea2b0b2b26a0d0fc5
--- /dev/null
+++ b/syscalldb.h
@@ -0,0 +1,45 @@
+#ifndef _SYSCALLDB_H_
+#define _SYSCALLDB_H_
+
+#ifdef __ASSEMBLER__
+
+#include "shim_entry_offsets.h"
+
+.macro SYSCALLDB
+leaq .Lafter_syscalldb\@(%rip), %rcx
+jmpq *%gs:SHIM_SYSCALLDB_OFFSET
+.Lafter_syscalldb\@:
+.endm
+
+#else /* !__ASSEMBLER__ */
+
+#include "shim_entry_offsets.h"
+
+#define XSTR(s) STR(s)
+#define STR(s) #s
+
+asm (
+    ".weak syscalldb\n"
+    ".type syscalldb, @function\n"
+    ".macro SYSCALLDB\n"
+    "leaq .Lafter_syscalldb\\@(%rip), %rcx\n"
+    "jmpq *%gs:" XSTR(SHIM_SYSCALLDB_OFFSET) "\n"
+    ".Lafter_syscalldb\\@:\n"
+    ".endm\n"
+);
+
+#define SYSCALLDB "SYSCALLDB\n"
+
+static inline int register_library(const char* name, unsigned long load_address)
+{
+    int (*shim_register_library)(const char*, unsigned long);
+    __asm__("movq %%gs:%c1, %0"
+            : "=r"(shim_register_library)
+            : "i"(SHIM_REGISTER_LIBRARY_OFFSET)
+            : "memory");
+    return shim_register_library(name, load_address);
+}
+
+#endif /* __ASSEMBLER__ */
+
+#endif /* _SYSCALLDB_H_ */
