--- jdk/make/sun/Makefile	2015-07-14 17:43:29.000000000 -0400
+++ jdk.new/make/sun/Makefile	2015-10-31 16:30:15.837466876 -0400
@@ -85,11 +85,17 @@ else
   DISPLAY_LIBS = awt splashscreen $(XAWT_SUBDIR) $(HEADLESS_SUBDIR)
   DISPLAY_TOOLS = applet
 endif
-SUBDIRS_desktop    = audio $(RENDER_SUBDIR) image \
-                     $(LWAWT_PRE_SUBDIR) $(DISPLAY_LIBS) $(DGA_SUBDIR) $(LWAWT_SUBDIR) \
-                     jawt font jpeg cmm $(DISPLAY_TOOLS) beans
-SUBDIRS_management = management
-SUBDIRS_misc       = $(ORG_SUBDIR) rmi $(JDBC_SUBDIR) tracing
+
+#SUBDIRS_desktop    = audio $(RENDER_SUBDIR) image \
+#                     $(LWAWT_PRE_SUBDIR) $(DISPLAY_LIBS) $(DGA_SUBDIR) $(LWAWT_SUBDIR) \
+#                     jawt font jpeg cmm $(DISPLAY_TOOLS) beans
+#SUBDIRS_management = management
+#SUBDIRS_misc       = $(ORG_SUBDIR) rmi $(JDBC_SUBDIR) tracing
+#SUBDIRS_tools      = native2ascii serialver tools jconsole
+
+SUBDIRS_desktop    =
+SUBDIRS_management =
+SUBDIRS_misc       = rmi tracing
 SUBDIRS_tools      = native2ascii serialver tools jconsole
 
 ifndef OPENJDK
Only in openjdk-7/jdk/src/share/lib/security: nss.cfg
